### DISCUSSION TITLE
Cannot start rocket from module inside sub folder (src/<dir0>

### DISCUSSION THREAD

**Discussion by astrometrics** [rwf2/Rocket#2481] [Questions]
Hi, thanks for Rocket

I'm a newb to rust, but I tried everything and I keep getting
error[E0468]: an `extern crate` loading macros must be at the crate root (for #[macro_use] extern crate rocket;)
also I cannot use the use directive to import... all just because I'm not in <project>/src, but <project>/src/<dir0>... That shouldn't be this hard... Can you please help me with that?

thanks

**Comment by robertwayne** [comment]
Can you share your directory / file structure, and what code you've tried?

Rocket can be started wherever you want in a handful of ways, depending on your needs. This is a minimal working example of starting in a sub-directory, however.

```
my_project/
src/
├─ main.rs
├─ subdir/
│  ├─ mod.rs
│  ├─ server.rs
Cargo.toml
```

```rust
// main.rs
#[macro_use]
extern crate rocket;

mod subdir;

use crate::subdir::server;

fn main() {
    server::main();
}
```

```rust
// subdir/server.rs
#[launch]
pub fn just_a_fn() -> _ {
    rocket::build().mount("/", routes![index])
}

#[get("/<_..>")]
async fn index() -> &'static str {
    "Hello, world!"
}
```

**Reply by astrometrics** [reply]
Hi @robertwayne , thanks for your answer. Your interpretation was all correct, with the exception that I want the starting point not in src/main, but in src/subdir/server.rs... Why is that even tough to do and not a trivial thing?
thanks again.

**Reply by robertwayne** [reply]
So are you wishing for your whole program entry point to be in the sub-directory?

In that case, you'd want to define it as such in your Cargo.toml:

```toml
# Cargo.toml
[[bin]]
name = "my_project"
path = "src/subdir/server.rs"
```

This also allows for multiple entry points. Then the example `subdir/server.rs` from above will act as your entry point and you could remove `main.rs`, or use it for other purposes.

If I'm misunderstanding the structure you want, could you show me what kind of file structure you're trying to make work?

**Reply by astrometrics** [reply]
Hi @robertwayne,
Cargo.toml was already configured that way... I get the error:
`error[E0468]: an `extern crate` loading macros must be at the crate root
 --> gdir/src/http/hsvr.rs:3:14
  |
3 | #[macro_use] extern crate rocket;
  |              ^^^^^^^^^^^^^^^^^^^^`

dependency is in place in Cargo.toml:
`[dependencies]

rocket  = "0.5.0-rc.2"`

Just for a test, commented everything in src/subdir/server.rs, and src/main.rs with Rocket initialization is working... So the problem continues...
thanks

**Reply by robertwayne** [reply]
That sounds like you're adding  `mod subdir` in `main.rs`, which would mean that *that* file is your designated top-level, so the `#[macro_use]` would have to go there. You can still use `subdir/server.rs` as the entry point and manually import stuff like `use rocket::launch`.

What exactly is in your `main.rs`, and what is your goal?

**Reply by astrometrics** [reply]
my src/main.rs is empty, all commented out. The goal now is to create a number of functionalities in a project with multiple points of entry in multiple subdirs. If all goes well then I will try to create a simple web server with some functions and other servers of other types in other subdirs (by the way is there a proper name for these, just like packages for java?). These servers will share functionalities available in some specific "packages" (subdirs).

**Reply by robertwayne** [reply]
> (by the way is there a proper name for these, just like packages for java?)

**Crate**: any stand-alone piece of code Rust can compile
**Package**: one more more crates
**Module**: collection of items _(fns, structs, etc)_; can be separate files or many per file

-----

We have to define our  `#[macro_use]` at the **top-level of a crate**. There can **only ever be one top-level of a crate**. Our **crate can only have one entry point**.

A common way to handling this would be to have all your shared functionality in a __library crate__, and you could have multiple stand-alone binaries alongside them; or one binary which uses CLI args to call different library code.

-----

It is **important to remember** that if you ever export your current module via `mod.rs` and then import that module via `mod my_module`, it is no longer the top-level of that crate, so you MUST move `#[macro_use]`  up the chain.

-----

I have tried annotating where package, crate, and module boundaries are in these trees, hopefully it isn't too messy.

```
my_package/
├── Cargo.toml <-- this marks our package root
└── src/ <-- CRATE
    ├── lib.rs <- MODULE
    └── my_crate/ <-- CRATE
    │   └── main.rs <-- MODULE, can use `#[macro_use]`  here as it is the top-level of our crate
    └── my_other_crate/ <-- CRATE
        └── main.rs <-- MODULE, can use `#[macro_use]`  here as it is the top-level of our crate
```

You could also move all of your binaries to the top level _(outside of `/src`)_:

```
my_package/
├── Cargo.toml  <-- this marks our package root
├── src/ <-- CRATE
│   └── lib.rs <- MODULE
└── my_crate/ <-- CRATE
    └── main.rs <-- MODULE, can use `#[macro_use]`  here as it is the top-level of our crate
```

In both of these examples, **you could alternatively define `#[macro_use]` in the `lib.rs` instead**, and then it would be available from all your other crates - whether they are sub-directories or outside of `src/`.

You could put a **single Rocket web server in your library and start it from each binary**, or you could put **separate Rocket web servers in each binary** and start them on their own. That's totally up to you, but both options work. It sounds like you want the latter.

If you still can't get it working, it would be easier if you could create a demo repository with your desired structure, and I can point out what needs to be changed to make it work. Unfortunately, without knowing very specifically what you want your code to look like, it's hard to give you a perfect answer!

Hopefully this helps guide you into the right direction for what you want, though! Rusts module system confused me a lot when I was learning the language.

**Reply by astrometrics** [reply]
Hi @robertwayne , thanks for the comprehensive response. #[macro_use] didnt work for me in any combination. But what surprisingly worked, despite VSCode marking everything as an error,  was simply
`
use rocket::{get, routes, launch, Rocket, Build};
`
that compiled and ran successfully and it's simple to understand. Also that solution does not require any src/main.rs, src/lib.rs or src/somedir/mod.rs gimmicks.
